src/dataset/pinecone_db_handler.py

```diff
@@ -0,0 +1,100 @@
+# dataset/pinecone_db_handler.py
+
+import os
+import json
+import pinecone
+
+from .settings import PINECONE_INDEX_NAME, PINECONE_VALUES_DIMS, PINECONE_METRIC, PINECONE_METADATA_ENTRIES, PINECONE_API_KEY, PINECONE_ENVIRONMENT
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+class PineconeDB:
+    def __init__(
+        self, 
+        create_index: bool = False,
+    ):
+        self.index_name = PINECONE_INDEX_NAME
+        
+        pinecone.init(
+            api_key = PINECONE_API_KEY,
+            environment = PINECONE_ENVIRONMENT,
+        )
+        
+        if create_index:
+            self.create_index()
+        
+        self.index = pinecone.Index(index_name=self.index_name)
+    
+    def __str__(self) -> str:
+        index_stats_response = self.index.describe_index_stats()
+        return f"{self.index_name}:\n{json.dumps(index_stats_response, indent=4)}"
+    
+    def upsert_entry(self, entry, chunks, embeddings, upsert_size=100):
+        self.index.upsert(
+            vectors=list(
+                zip(
+                    [f"{entry['id']}_{str(i).zfill(6)}" for i in range(len(chunks))], 
+                    embeddings.tolist(), 
+                    [
+                        {
+                            'entry_id': entry['id'],
+                            'source': entry['source'],
+                            'title': entry['title'],
+                            'authors': entry['authors'],
+                            'text': chunk,
+                        } for chunk in chunks
+                    ]
+                )
+            ),
+            batch_size=upsert_size
+        )
+    
+    def upsert_entries(self, entries_batch, chunks_batch, chunks_ids_batch, embeddings, upsert_size=100):
```

Comment:
isn't this pretty much `upsert_entry` with an extra loop? i.e. it might be worth either calling `upsert_entry` from this function, or abstracting it away so that there is an additional function that does the actual work, which is called by these functions

Reply:
Yes, but the reason it's faster is that it makes fewer index.upsert upsert calls, which used to be one of the major speed bottlenecks. It now calls index.upsert once per CHUNK_SIZE chunks instead of once per entry. Calling index.upsert with 10 vectors 100 times is much slower than calling it once with 1,000 vectors, which is the reason this whole batching endeavor was worth implementing in the first place. My (admittedly limitted) testing seems to show it's much faster than it used to be. It still takes over two hours to update the full dataset*, but when we'll run src/main.py just to update the current dataset with the past week's posts/papers/etc, it will only take a few seconds/minutes.
There are certainly ways to make the code cleaner, but I don't know of a simple way to make it cleaner while not making it slower. The process of creating the 'vectors' variable could be done in another method, I suppose.

*on colab with a GPU. Parenthetically, I tested it there because I wanted to use the free sentence transformer for testing and running it on my computer would have taken hundreds of hours. It also seems like Pinecone upserts are much faster on colab than on my computer. It's worth noting that the Pinecone documentation mentions this is normal and that google colab/google cloud/azure cloud/amazong cloud are much faster than home networks for pinecone upserts. I don't know if this means we'll essentially be forced to update the dataset on these cloud servers, but in any case I have a google colab that gets the most recent version of the right stampy-chat branch from the github repo and updates the pinecone dataset from there. One issue to fix however is to store the sql database externally, possibly on google drive, to not have to start from scratch every time.

Reply:
No, you're totally correct on that front, I was thinking more a step earlier, eg.
```
def upsert(self, ids, embeddings, chunks, upsert_size):
    data = [
                        {
                            'entry_id': entry['id'],
                            'source': entry['source'],
                            'title': entry['title'],
                            'authors': entry['authors'],
                            'text': chunk,
                        } for chunk in chunks
    ] 
    return self.index.upsert(
        vectors=list(zip(ids, embeddings.tolist(), data)), batch_size=upsert_size
   )

def upsert_entry(self, entry, chunks, embeddings, upsert_size=100):
    self.upsert(
        [f"{entry['id']}_{str(i).zfill(6)}" for i in range(len(chunks))],
        embeddings,
        chunks
       upsert_size
   )

 def upsert_entries(self, entries_batch, chunks_batch, chunks_ids_batch, embeddings, upsert_size=100):
    self.upsert(chunks_ids_batch, embeddings, (chunk for entry in entries_batch for chunk in chunks_batch), upsert_size)